tests/integration/test_integration.py

```diff
@@ -29,209 +31,351 @@
 SELF_SIGNED_CERTIFICATES_APPLICATION_NAME = "self-signed-certificates"
 VAULT_KV_REQUIRER_APPLICATION_NAME = "vault-kv-requirer"
 VAULT_PKI_REQUIRER_APPLICATION_NAME = "tls-certificates-requirer"
+S3_INTEGRATOR_APPLICATION_NAME = "s3-integrator"
+MINIO_APPLICATION_NAME = "minio"
 
 VAULT_KV_LIB_DIR = "lib/charms/vault_k8s/v0/vault_kv.py"
 VAULT_KV_REQUIRER_CHARM_DIR = "tests/integration/vault_kv_requirer_operator"
 
-NUM_VAULT_UNITS = 4
-
-k8s = KubernetesClient()
+MINIO_S3_ACCESS_KEY = "minio_access_key"
+MINIO_S3_SECRET_KEY = "minio_secret_key"
+MINIO_CONFIG = {
+    "access-key": MINIO_S3_ACCESS_KEY,
+    "secret-key": MINIO_S3_SECRET_KEY,
+}
 
+NUM_VAULT_UNITS = 3
 
-def copy_lib_content() -> None:
-    shutil.copyfile(src=VAULT_KV_LIB_DIR, dst=f"{VAULT_KV_REQUIRER_CHARM_DIR}/{VAULT_KV_LIB_DIR}")
+k8s = KubernetesClient()
 
 
 class TestVaultK8s:
-    @staticmethod
-    async def deploy_charm(ops_test: OpsTest, charm: Path) -> None:
-        """Deploys charm.
-
-        Args:
-            ops_test: Ops test Framework.
-            charm: Charm path.
-        """
+    @pytest.fixture(scope="module")
+    async def deploy_phase_1(self, ops_test: OpsTest):
```

Comment:
We introduce the concept of a "phase" here but this is not described anywhere. What is a phase? Do we really need this concept? I understand why we want to deploy everything in parallel but we can have this without the concept of a "phase"

Reply:
1) A phase of testing is simply a grouping of tests, grouped around testing the behavior of a couple of selected charms which the tests will interact with. Once a phase is over (all of the tests that call the fixture complete), those applications are removed.

2) The main reasons having phases is convenient are:
Save memory in the github runner so we don't have pods crashing and making the tests unreliable.
Split up the testing suite of charms by a logical grouping that is explicitly called out to avoid confusion.
Still parallelize as much of the building and deployment process as we can.

I'm open to renaming them or coming up with a better solution with the goal of: 
* reducing the total time it takes for the tests to complete
* managing the memory of the runner

Where should I place this description in the code?

Reply:
I'm not convinced we need the concept of phases, let's name the methods based on what they actually do.

Reply:
Also, please add some docstrings to the new methods like this one.

Reply:
Without phases, the deployments of charms will no longer be parallel. We can't deploy every dependency together because it will crash the runner.

Other ideas are:
I can make a fixture for each of the charms and have them deploy serially, which will take much longer to deploy.
I can just create two files called integration_test_part_1.py and integration_test_part_2.py and copy and paste the duplicate code that will create.

Reply:
I think the phases split might not be a perfect fit here, and maybe things should be split in more fixtures, but I am not sure how to do that and keep most of the async stuff. I find mostly confusing the part where some tests depend on `phase2`, and that seems to also implicitly depend on parts of `phase1`. The cleanup of `phase1` is thus incomplete and left to the framework.

Reply:
Huh, I just noticed that Traefik isn't being torn down and at the same time I mis-marked it as phase 2, and the self signed cert is also in phase 2, so the 3 bugs eliminated each other to make this thing work. That's fixed now. 

I'm not a big fan of just calling these phases but considering the alternatives I provided above, I'd still prefer going with phases for now because performance is more of a concern when testing Vault features. If you guys prefer another option above or a better paradigm, I'm open to changing it.

Reply:
I split the tests into 3 different classes with parts, and made vault a module level fixture, and made the deployments class level. The naming is still iffy, but I think it reads much better this way and is way more pytestic or something like that.